FunctionBar.c

```diff
@@ -92,6 +92,14 @@ int FunctionBar_draw(const FunctionBar* this) {
    return FunctionBar_drawExtra(this, NULL, -1, false);
 }
 
+static size_t FunctionBar_safeLen(const char *s) {
+   size_t slen = mbstowcs(NULL, s, 0);
+   if (slen == (size_t)-1) {
+      slen = strlen(s);
```

Comment:
I would like some explanation on the logic. `mbstowcs` returns `(size_t)-1` when encountered an illegal byte, but why fall back to `strlen`? How is this function meant to be used?

Reply:
This function implements "get multi-byte string length, otherwise byte length". It is used below to position the cursor on the filter input field. In the happy path the string will have valid encoding, `mbstowcs` will succeed, and we'll position the cursor in the right place. If `s` is garbage, byte length is the only sensible fallback. `safeLen` is maybe not the best name.

Thanks for the review. I'll add your fixes shortly.

Reply:
@adse So this function is meant to be getting a "display width"?
Unfortunately there is a complexity here. Not every Unicode character occupies one column in the terminal. CJK characters would occupy two. So assuming the "display width" of the string is the same as number of code points in the string (i.e. the return value of `mbstowcs`) is wrong. The best approach as far as I can find is to use [`wcwidth`](https://man7.org/linux/man-pages/man3/wcwidth.3.html) and [`wcswidth`](https://man7.org/linux/man-pages/man3/wcswidth.3.html).

Reply:
You're right. (I wrote this in the PR description.) This was only meant as an incremental improvement with proper handling and rendering as a follow-up.

Reply:
I renamed `FunctionBar_safeLen` to `FunctionBar_displayWidth` which sums `wcwidth` of each wchar and falls back on `strlen` on invalid encoding.

Reply:
I need some time to review this, as I don't think falling back to `strlen` is a good idea. Ideally the string passed to `FunctionBar_displayWidth` should have been sanitized so that no invalid byte would exist (if there were invalid bytes or by sequences they should be replaced with U+FFFD as part of the sanitization). The other issue I can think of is that `wcwidth` might return -1 for non-printable characters and you need to handle that as a special case (maybe treat them as zero width?).

Reply:
Is sanitizing the process name the right thing to do? The kernel doesn't put any encoding restrictions on process names, so users should ideally be able to filter by arbitrary binary as well.

For that reason, I'd vote for not sanitizing filter strings, and rendering anything non-printable or invalidly encoded as "?" (or U+FFFD). htop seems to do this already when printing comm lines. (Try `htop  -F $'\t'`.) `top` and `ps` both do something similar to what I'm describing.

`htop  -F $'\n'` is presently buggy. This will fix that as well.

I can implement this if people agree. It'll require changing some more of `FunctionBar.c`.

Reply:
@adsr By "sanitizing" I mean filter out any Unicode control character that may come from user input, converting them into U+FFFD characters _or_ print them out in some form of escape sequences. You do need to filter that for display purposes (something that htop didn't implement yet, unfortunately). Otherwise you could see a spoofing attack like in [this article](https://www.vadesecure.com/en/blog/how-hackers-are-using-a-20-year-old-text-trick-to-phish-microsoft-365-users). I'm not sure if the Filter function of htop is going to support escape sequences, but accepting user input (from `getch` or `get_wch`) without some kind of checking is generally a bad idea.

Specifically, at least, never allow non-printable characters from `get_wch` input:
```
wint_t wkey;
int wchrv = get_wch(&wkey);
if (wchrv == OK) {
   if (!iswprint(wkey)) {
      // Disallow. Or convert this to U+FFFD
      wkey = (wint_t )0xFFFD;
   }
}
```

Reply:
> You do need to filter that for display purposes (something that htop didn't implement yet, unfortunately).

I agree that filtering on display is needed, however I disagree with your code snippet (filtering input). As I stated, non-printables and invalid encodings are allowed in process names, so disallowing them as input would prevent users from matching such processes. Of course this is a very narrow edge case and naturally I will defer to the maintainers, but that is my view.

@BenBE or someone else want to weigh in?

Reply:
> I agree that filtering on display is needed, however I disagree with your code snippet (filtering input). As I stated, non-printables and invalid encodings are allowed in process names, so disallowing them as input would prevent users from matching such processes. Of course this is a very narrow edge case and naturally I will defer to the maintainers, but that is my view.

My idea is that the invalid characters are disallowed, in raw, from user input. They may only be input via escape sequences.

Reply:
@Explorer09 Can you elaborate on the "input via escape sequences" part? Currently the filter/search functions don't do any escape sequence processing. If this was to be added, that'd be a different feature.

Also on the part of invalid encodings: As those are possible to occur as part of any process name/information we should not hinder filtering/searching for them. Thus if you limit inputs to valid character sequences only you'll also have to allow the user to specify if they wanted to look for an invalid encoding (intentionally).

Reply:
@BenBE An example is embedded newlines in filenames, if you are adding that to the filter string, you should probably type it as `\n` and not the Enter key as the latter would mean something else. Likewise, you will type `\033` for the ASCII ESC character in the filename and not the Escape key. That's the filtering I was talking about.

You will need to filter this from the keyboard input (within the GUI), but not from the `-F` option argument (the escape sequence of the latter is processed by the shell).

Reply:
@Explorer09 Exactly. And this handling is currently not implemented. Same for when you need to look for `Hallö Wörld` in a process that uses ISO-8859-1 instead of UTF-8 …